src/algebra/module/projective.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2021 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+
+import algebra.module.basic
+import linear_algebra.finsupp
+import linear_algebra.basis
+
+/-!
+
+# Projective modules
+
+This file contains a definition of a projective module, the proof that
+our definition is equivalent to a lifting property, and the
+proof that all free modules are projective.
+
+## Main definitions
+
+Let `R` be a ring (or a semiring) and let `M` be an `R`-module (or a semimodule).
+
+* `is_projective R M` : the proposition saying that `M` is a projective `R`-module.
+
+## Main theorems
+
+* `is_projective.lifting_property` : a map from a projective module can be lifted along
+  a surjection.
+
+* `is_projective.of_lifting_property` : If for all R-module surjections `A →ₗ B`, all
+  maps `M →ₗ B` lift to `M →ₗ A`, then `M` is projective.
+
+* `is_projective.of_free` : Free modules are projective
+
+## Implementation notes
+
+The actual definition of projective we use is that the natural R-module map
+from the free R-module on the type M down to M splits. This is more convenient
+than certain other definitions which involve quantifying over universes,
+and also universe-polymorphic (the ring and module can be in different universes).
+
+Everything works for semirings and semimodules except that apparently
+we don't have free semimodules, so here we stick to rings.
+
+`is_projective.of_lifting_property` is not universe polymorphic.
+
+## References
+
+https://en.wikipedia.org/wiki/Projective_module
+
+## TODO
+
+Direct sum of two projective modules is projective. Arbitrary
+sum of projective modules is projective. Any module admits
+a surjection from a projective module. All of these should
+be relatively straightforward.
+
+## Tags
+
+projective module
+
+-/
+
+universes u v
+
+/- The actual implementation we choose: `M` is projective if the natural surjection
+   from the free `R`-module on `M` to `M` splits. -/
+/-- An R-module is projective if it is a direct summand of a free module, or equivalently
+  if maps from the module lift along surjections. There are several other equivalent
+  definitions. -/
+def is_projective
+  (R : Type u) [semiring R] (M : Type v) [add_comm_monoid M] [semimodule R M] : Prop :=
+∃ s : M →ₗ[R] (M →₀ R), function.left_inverse (finsupp.total M M R id) s
+
+namespace is_projective
+
+section semiring
+
+variables {R : Type u} [semiring R] {M : Type v} [add_comm_monoid M] [semimodule R M]
+  {A : Type*} [add_comm_group A] [semimodule R A] {B : Type*} [add_comm_group B] [semimodule R B]
```

Comment:
Does it make sense to work with `M_1` and `M_2`? I keep finding myself thinking that `B` is an `A`-algebra.

Reply:
For me, the standard terminology would be to call the projective module P and the surjection M - > N.

Reply:
I like Julian's idea best. I'll change it!

Reply:
Sorry, this fell off my radar for a bit. I've changed all the variables so the projective module is P and the surjection is M -> N.